x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StDistance.java

```diff
@@ -0,0 +1,264 @@
+import org.elasticsearch.compute.operator.EvalOperator;
+import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.geometry.Point;
+import org.elasticsearch.lucene.spatial.CoordinateEncoder;
+import org.elasticsearch.xpack.esql.EsqlIllegalArgumentException;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.core.util.SpatialCoordinateTypes;
+import org.elasticsearch.xpack.esql.evaluator.mapper.EvaluatorMapper;
+import org.elasticsearch.xpack.esql.expression.function.Example;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+
+import java.io.IOException;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.esql.core.type.DataType.DOUBLE;
+import static org.elasticsearch.xpack.esql.expression.function.scalar.spatial.SpatialRelatesUtils.makeGeometryFromLiteral;
+
+/**
+ * Computes the distance between two points.
+ * For cartesian geometries, this is the pythagorean distance in the same units as the original coordinates.
+ * For geographic geometries, this is the circular distance along the great circle in meters.
+ * The function `st_distance` is defined in the <a href="https://www.ogc.org/standard/sfs/">OGC Simple Feature Access</a> standard.
+ * Alternatively it is described in PostGIS documentation at <a href="https://postgis.net/docs/ST_Distance.html">PostGIS:ST_Distance</a>.
+ */
+public class StDistance extends BinarySpatialFunction implements EvaluatorMapper {
+    // public for test access with reflection
+    public static final DistanceCalculator GEO = new GeoDistanceCalculator();
+    // public for test access with reflection
+    public static final DistanceCalculator CARTESIAN = new CartesianDistanceCalculator();
+
+    protected static class GeoDistanceCalculator extends DistanceCalculator {
+        protected GeoDistanceCalculator() {
+            super(SpatialCoordinateTypes.GEO, CoordinateEncoder.GEO);
+        }
+
+        @Override
+        protected double distance(Point left, Point right) {
+            // TODO: investigate if we need to use the more complex validation in Lucenes Circle2D::HaversinDistance class
+            return SloppyMath.haversinMeters(left.getY(), left.getX(), right.getY(), right.getX());
+        }
+    }
+
+    protected static class CartesianDistanceCalculator extends DistanceCalculator {
+
+        protected CartesianDistanceCalculator() {
+            super(SpatialCoordinateTypes.CARTESIAN, CoordinateEncoder.CARTESIAN);
+        }
+
+        @Override
+        protected double distance(Point left, Point right) {
+            return Math.sqrt(Math.pow(left.getX() - right.getX(), 2) + Math.pow(left.getY() - right.getY(), 2));
```

Comment:
I think for squares, it is an overkill to call `Math.pow`

```suggestion
           final double diffX = left.getX() - right.getX();
           final double diffY = left.getY() - right.getY(); 
           return Math.sqrt(diffX * diffX + diffY * diffY);
```